modules/table/src/main/java/org/apache/ignite/internal/table/distributed/TableManager.java

```diff
@@ -1317,152 +1379,153 @@ private RuntimeException convertThrowable(Throwable th) {
 
-        var newBaseline = currClusterMembers
-                .stream().filter(n -> nodes.contains(n.name())).collect(Collectors.toSet());
+                String grpId = partitionRaftGroupName(tblId, part);
 
-        updateAssignments(currClusterMembers);
+                Supplier<RaftGroupListener> raftGrpLsnrSupplier = () -> new PartitionListener(tblId,
+                        new VersionedRowStore(
+                                tablesByIdVv.latest().get(tblId).internalTable().storage().getOrCreatePartition(part), txManager));
 
-        if (!newBaseline.equals(currClusterMembers)) {
-            updateAssignments(newBaseline);
-        }
-    }
+                Supplier<RaftGroupEventsListener> raftGrpEvtsLsnrSupplier = () -> raftGroupEventsListener(
+                        tablesByIdVv.latest().get(tblId).name(),
+                        part,
+                        grpId);
 
-    /**
-     * Update assignments for all current tables according to input nodes list. These approach has known issues {@link
-     * Ignite#setBaseline(Set)}.
-     *
-     * @param clusterNodes Set of nodes for assignment.
-     */
-    private void updateAssignments(Set<ClusterNode> clusterNodes) {
-        var setBaselineFut = new CompletableFuture<>();
 
-        var changePeersQueue = new ArrayList<Supplier<CompletableFuture<Void>>>();
+                List<List<ClusterNode>> assignments = (List<List<ClusterNode>>)
+                        ByteUtils.fromBytes(((ExtendedTableConfiguration) tablesCfg.tables().get(tbl.name())).assignments().value());
 
-        tablesCfg.tables()
-                .change(tbls -> {
-                    changePeersQueue.clear();
+                List<ClusterNode> newPeers = ((List<ClusterNode>) ByteUtils.fromBytes(evt.entryEvent().newEntry().value()));
 
-                    for (int i = 0; i < tbls.size(); i++) {
-                        tbls.createOrUpdate(tbls.get(i).name(), changeX -> {
-                            ExtendedTableChange change = (ExtendedTableChange) changeX;
-                            byte[] currAssignments = change.assignments();
+                RaftGroupService raftGrpSvc = null;
 
-                            List<List<ClusterNode>> recalculatedAssignments = AffinityUtils.calculateAssignments(
-                                    clusterNodes,
-                                    change.partitions(),
-                                    change.replicas());
+                try {
+                    var deltaPeers = newPeers.stream()
+                            .filter(p -> !assignments.get(part).contains(p))
+                            .collect(Collectors.toList());
 
-                            if (!recalculatedAssignments.equals(ByteUtils.fromBytes(currAssignments))) {
-                                change.changeAssignments(ByteUtils.toBytes(recalculatedAssignments));
+                    raftGrpSvc = raftMgr.updateRaftGroup(grpId, assignments.get(part), deltaPeers, raftGrpLsnrSupplier,
```

Comment:
Here is excess invocation, because we do nothing with peers in RaftGroupSercive, although recreate them (I know it is to start raft server if it is required, but maybe we can avoid of restarts of RaftGroupSercive).

Reply:
> Here is excess invocation, because we do nothing with peers in RaftGroupSercive, although recreate them (I know it is to start raft server if it is required, but maybe we can avoid of restarts of RaftGroupSercive).

Yep, I believe that, we should update internalTableImpl with new raft group services.

Reply:
Guys, we must update raft group services, only after successful rebalance and configuration updates, but not in the listener of pending assignments - it's too early, this rebalance can be failed.

But, I agree, that we don't need to recreate the raft services here, I changed this code to:
- on the non-leader nodes just start new raft group nodes
- on the leader node run changePeersAsync on the already created raft group service from internalTable